Add tracing support for optional Device and Layout

izdeby opened the pull request:
This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add tracing support for optional Device and Layout types.

--------------

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30983 Updated codegen to remove TensorOptions from generated code
* #30982 Prepare templates
* #30981 Updated DispatchKeyExtractor to expect TensorOptions
* #30980 Added C++ API test
* **#30979 Add tracing support for optional Device and Layout**

Differential Revision: [D18912685](https://our.internmc.facebook.com/intern/diff/D18912685)

izdeby committed:
Add tracing support for optional Device and Layout

[ghstack-poisoned]

gchanan commented:
Someone from JIT (e.g. @eellison should review this).

eellison commented on torch/csrc/jit/tracer.cpp:
```diff
@@ -504,6 +504,30 @@ void addInputs(
+  if (value.has_value()) {
+    detail::genericAddInput(n, static_cast<int64_t>(*value));
+  } else {
+    Graph* g = n->owningGraph();
```
It would be nice if you factored the 8 copy-pasta's of this three-line block into it's own function (i am one of those copy-paster's...)

eellison reviewed (approved):
LGTM. I don't know how exactly you plan on landing this stack; could you not land this PR without adding tests for these changes ? If it's in an upcoming PR that's fine.

izdeby committed:
Update on "Add tracing support for optional Device and Layout"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add tracing support for optional Device and Layout types.

--------------




[ghstack-poisoned]

izdeby committed:
Update on "Add tracing support for optional Device and Layout"

This stack is a first step toward an effort to fix, clean up and simplify code generation logic.  Please see the master [task](https://github.com/pytorch/pytorch/issues/30405) to see related discussions and all the known issues.

Main focus of these changes is TensorOptions in code generation. 
Goals:
- Remove TensorOptions from generated code wherever it's possible. Leave it only in python/C++ API layers.
- Refactor TensorOptions logic to a single place.
- Log all discovered issues.

Non goals:
- Fix Everything! 
- Remove all the hacks in code generation scripts.
- Clean up and defector all code generation scripts.

--------------
In this PR: 
Add tracing support for optional Device and Layout types.

--------------


Differential Revision: [D18912685](https://our.internmc.facebook.com/intern/diff/D18912685)

[ghstack-poisoned]
